### PR TITLE
swap orddict for maps

### PR DESCRIPTION
**Pull request by dch**
Thanks Kansi! LGTM overall, nice work.

This needs a rebase after I added the ppspp_options_SUITE tests this morning. Good timing depending on your point of view!

Some general comments:
- these `%% @type options() = {options, options_dict()}.` need to stay duplicated, edoc/edown needs them for generating typespec info in the docs
- please remove the commented-out orddict lines
- where we are creating the option for the first time we should fail if the key already exists
- ensure the function body comments are updated to match the changed code :-)

Before final push, we should squash this down to 1 commit, and confirm with `make reindent` that all is tidy. Your indentation (perhaps emacs?) is different to vim's one.

### CODE REVIEW

File: src/ppspp_datagram.erl [hunk 1]
```diff
@@ -33,14 +33,25 @@
...
+-opaque endpoint()        :: #{}.
+-opaque endpoint_option() :: #{ip => inet:ip_address()}
+                           | #{socket => inet:socket()}
+                           | #{port => inet:port_number()}
+                           | #{transport => udp}
+                           | #{uri => string()}
+                           | ppspp_channel:channel().
+
+
+-opaque datagram()        :: {datagram, map(endpoint() | ppspp_message:messages())}.
```
Comment (elvisci) [inline src/ppspp_datagram.erl]: Line 53 is too long: -opaque datagram()        :: {datagram, map(endpoint() | ppspp_message:messages())}..

File: src/ppspp_have.erl [hunk 2]
```diff
@@ -48,7 +48,7 @@
 
 unpack(Addressing_Method, Message) ->
     {Chunk_Spec, Maybe_Messages} = ppspp_chunk:unpack(Addressing_Method, Message),
```
Comment (elvisci) [inline src/ppspp_have.erl]: Line 50 is too long:     {Chunk_Spec, Maybe_Messages} = ppspp_chunk:unpack(Addressing_Method, Message),.

File: src/ppspp_options.erl [hunk 3]
```diff
@@ -322,13 +340,20 @@ pack(_) -> <<>>.
...
-                 {version,1}]},
+    Expected = #{chunk_addressing_method        => chunking_32bit_chunks,
+                 chunk_size                     => 1024,
+                 content_integrity_check_method => merkle_hash_tree,
+                 merkle_hash_tree_function      => sha,
+                 minimum_version                => 1,
+                 swarm_id                       => Swarm_id,
+                 version                        => 1},
+    %% Expected = {options,
+    %%             [{chunk_addressing_method,chunking_32bit_chunks},
```
Comment (elvisci) [inline src/ppspp_options.erl]: Missing space after "," on line 351

File: src/ppspp_options.erl [hunk 4]
```diff
@@ -322,13 +340,20 @@ pack(_) -> <<>>.
...
+    %%              {chunk_size,1024},
```
Comment (elvisci) [inline src/ppspp_options.erl]: Missing space after "," on line 352

File: src/ppspp_options.erl [hunk 5]
```diff
@@ -322,13 +340,20 @@ pack(_) -> <<>>.
...
+    %%              {content_integrity_check_method,merkle_hash_tree},
```
Comment (elvisci) [inline src/ppspp_options.erl]: Missing space after "," on line 353

File: src/ppspp_options.erl [hunk 6]
```diff
@@ -322,13 +340,20 @@ pack(_) -> <<>>.
...
+    %%              {merkle_hash_tree_function,sha},
```
Comment (elvisci) [inline src/ppspp_options.erl]: Missing space after "," on line 354

File: src/ppspp_options.erl [hunk 7]
```diff
@@ -322,13 +340,20 @@ pack(_) -> <<>>.
...
+    %%              {minimum_version,1},
```
Comment (elvisci) [inline src/ppspp_options.erl]: Missing space after "," on line 355

File: src/ppspp_options.erl [hunk 8]
```diff
@@ -322,13 +340,20 @@ pack(_) -> <<>>.
...
+    %%              {swarm_id, Swarm_id},
+    %%              {version,1}]},
```
Comment (elvisci) [inline src/ppspp_options.erl]: Missing space after "," on line 357

File: src/ppspp_options.erl [hunk 9]
```diff
@@ -81,44 +82,49 @@
...
+          Maybe_Options/binary >>, Options) ->
     Method = case Maybe_Method of
                  0 -> no_integrity_protection;
                  1 -> merkle_hash_tree;
                  2 -> sign_all;
                  3 -> unified_merkle_hash_tree
              end,
-    Options = orddict:store(content_integrity_check_method, Method, Options0),
-    unpack(Maybe_Options, Options);
+    %% Options = orddict:store(content_integrity_check_method, Method, Options0),
```
Comment (elvisci) [inline src/ppspp_options.erl]: Line 121 is too long:     %% Options = orddict:store(content_integrity_check_method, Method, Options0),.

File: src/ppspp_options.erl [hunk 10]
```diff
@@ -126,23 +132,25 @@ unpack( <<?PPSPP_MERKLE_HASH_FUNCTION,
...
+    %% Options = orddict:store(merkle_hash_function, Hash_Function, Options0),
+    unpack(Maybe_Options, Options#{merkle_hash_function => Hash_Function});
+
 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
 unpack( <<?PPSPP_LIVE_SIGNATURE_ALGORITHM,
           Maybe_Algorithm:?BYTE,
-          Maybe_Options/binary >>, Options0) ->
+          Maybe_Options/binary >>, Options) ->
     Algorithm = case Maybe_Algorithm of
                     %% TODO we can parse this yet but need normative algorithm names
```
Comment (elvisci) [inline src/ppspp_options.erl]: Line 143 is too long:                     %% TODO we can parse this yet but need normative algorithm names.

File: src/ppspp_options.erl [hunk 11]
```diff
@@ -173,20 +183,22 @@ unpack( <<?PPSPP_LIVE_DISCARD_WINDOW, Maybe_Discard_Window/binary>>,
                  16#ffffffffffffffff -> ppspp_live_discard_window_keep_all_chunks;
```
Comment (elvisci) [inline src/ppspp_options.erl]: Line 183 is too long:                  16#ffffffffffffffff -> ppspp_live_discard_window_keep_all_chunks;.

File: src/ppspp_options.erl [hunk 12]
```diff
@@ -173,20 +183,22 @@ unpack( <<?PPSPP_LIVE_DISCARD_WINDOW, Maybe_Discard_Window/binary>>,
...
+    unpack(Maybe_Options, Options#{ppspp_live_discard_window => Window});
+
 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
 unpack( <<?PPSPP_SUPPORTED_MESSAGES,
           Length:?BYTE,
           Maybe_Supported_Message_Types:Length/big,
-          Maybe_Options/binary >>, Options0) ->
+          Maybe_Options/binary >>, Options) ->
     Supported_Message_Types = case Maybe_Supported_Message_Types of
                                   %% TODO probably needs a custom bitmatching parser
```
Comment (elvisci) [inline src/ppspp_options.erl]: Line 195 is too long:                                   %% TODO probably needs a custom bitmatching parser.

File: src/ppspp_options.erl [hunk 13]
```diff
@@ -173,20 +183,22 @@ unpack( <<?PPSPP_LIVE_DISCARD_WINDOW, Maybe_Discard_Window/binary>>,
...
                                   _ -> ppspp_supported_messages_not_yet_implemented
```
Comment (elvisci) [inline src/ppspp_options.erl]: Line 196 is too long:                                   _ -> ppspp_supported_messages_not_yet_implemented.

File: src/ppspp_options.erl [hunk 14]
```diff
@@ -173,20 +183,22 @@ unpack( <<?PPSPP_LIVE_DISCARD_WINDOW, Maybe_Discard_Window/binary>>,
...
                               end,
-    Options = orddict:store(ppspp_supported_message_types,
-                            Supported_Message_Types, Options0),
-    unpack(Maybe_Options, Options);
+    %% Options = orddict:store(ppspp_supported_message_types,
+    %%                         Supported_Message_Types, Options0),
+    unpack(Maybe_Options, Options#{ppspp_supported_message_types => Supported_Message_Types});
```
Comment (elvisci) [inline src/ppspp_options.erl]: Line 200 is too long:     unpack(Maybe_Options, Options#{ppspp_supported_message_types => Supported_Message_Types});.

File: src/ppspp_options.erl [hunk 15]
```diff
@@ -228,8 +239,8 @@ get_chunk_addressing_method(Options) ->
 %% @end
 -spec get_content_integrity_check_method(options()) -> any().
 
-get_content_integrity_check_method(Options) ->
-    get(content_integrity_check_method, Options).
+get_content_integrity_check_method(#{content_integrity_check_method := Method} = _Options) ->
```
Comment (elvisci) [inline src/ppspp_options.erl]: Line 242 is too long: get_content_integrity_check_method(#{content_integrity_check_method := Method} = _Options) ->.

File: src/ppspp_options.erl [hunk 16]
```diff
@@ -228,8 +239,8 @@ get_chunk_addressing_method(Options) ->
...
+    Method.
 
 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
 %% @doc get_merkle_hash_function/1 returns the hash function for the provided swarm
```
Comment (elvisci) [inline src/ppspp_options.erl]: Line 246 is too long: %% @doc get_merkle_hash_function/1 returns the hash function for the provided swarm.

File: src/ppspp_options.erl [hunk 17]
```diff
@@ -237,8 +248,8 @@ get_content_integrity_check_method(Options) ->
 %% @end
 -spec get_merkle_hash_tree_function(options()) -> any().
 
-get_merkle_hash_tree_function(Options) ->
-    get(merkle_hash_tree_function, Options).
+get_merkle_hash_tree_function(#{merkle_hash_tree_function := HashFunc} = _Options) ->
```
Comment (elvisci) [inline src/ppspp_options.erl]: Line 251 is too long: get_merkle_hash_tree_function(#{merkle_hash_tree_function := HashFunc} = _Options) ->.

File: src/ppspp_options.erl [hunk 18]
```diff
@@ -237,8 +248,8 @@ get_content_integrity_check_method(Options) ->
...
+    HashFunc.
 
 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
 %% @doc get_minimum_version/1 returns the lowest accepted PPSP version for the swarm
```
Comment (elvisci) [inline src/ppspp_options.erl]: Line 255 is too long: %% @doc get_minimum_version/1 returns the lowest accepted PPSP version for the swarm.

File: src/ppspp_options.erl [hunk 19]
```diff
@@ -294,14 +305,21 @@ use_default_options() ->
...
+    %%  orddict:from_list( [{swarm_id, Swarm_id},
+    %%                      {chunk_addressing_method, chunking_32bit_chunks},
+    %%                      {chunk_size, ?PPSPP_DEFAULT_CHUNK_SIZE},
+    %%                      {content_integrity_check_method, merkle_hash_tree},
+    %%                      {merkle_hash_tree_function, sha},
+    %%                      {minimum_version, ?PPSPP_RFC_VERSION},
+    %%                      {version, ?PPSPP_RFC_VERSION}])}.
 
 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
 %% @doc pack an orddict of ppspp erlang terms into a binary PPSPP message segment
```
Comment (elvisci) [inline src/ppspp_options.erl]: Line 325 is too long: %% @doc pack an orddict of ppspp erlang terms into a binary PPSPP message segment.

File: test/ppspp_have_SUITE.erl [hunk 20]
```diff
@@ -35,5 +35,6 @@ have(_Config) ->
     Method = chunk_32bit_chunks,
     Garbage = crypto:strong_rand_bytes(10),
     Chunks = <<15:64, Garbage/binary>>,
-    Expected = {{have,{chunk_spec,{chunk_32bit_chunks,0,15}}}, Garbage},
+    %% Expected = {{have,{chunk_spec,{chunk_32bit_chunks,0,15}}}, Garbage},
+    Expected = { #{have => #{chunk_spec => {chunk_32bit_chunks,0,15} }}, Garbage},
```
Comment (elvisci) [inline test/ppspp_have_SUITE.erl]: Missing space after "," on line 39

Comment (elvisci) [inline test/ppspp_have_SUITE.erl]: Line 39 is too long:     Expected = { #{have => #{chunk_spec => {chunk_32bit_chunks,0,15} }}, Garbage},.

File: src/ppspp_channel.erl [hunk 21]
```diff
@@ -79,8 +80,8 @@ pack(_Message) -> <<>>.
 %% @end
 -spec acquire_channel(ppspp_options:swarm_id()) -> channel().
 acquire_channel(Swarm_id) ->
-    Channel = find_free_channel(Swarm_id, 0),
-    {channel, Channel}.
+    Channel = find_free_channel(Swarm_id, 1),
```
Comment (dch) [inline src/ppspp_channel.erl]: why does this need to change from 0 -> 1?

File: src/ppspp_chunk.erl [hunk 22]
```diff
@@ -36,22 +36,18 @@
...
-chunk_32bit_bins
-| chunk_64bit_bins
-| chunk_64bit_bytes
-| chunk_32bit_chunks
-| chunk_64bit_chunks.
+-type addressing_method() :: chunk_32bit_bins
+                           | chunk_64bit_bins
+                           | chunk_64bit_bytes
+                           | chunk_32bit_chunks
+                           | chunk_64bit_chunks.
```
Comment (dch) [inline src/ppspp_chunk.erl]: curious, is this indentation from emacs? or vim? or something else?

**Comment by dch** [comment]
@kansi ^^

File: src/ppspp_chunk.erl (hunk 22 above)
Comment (kansi) [inline src/ppspp_chunk.erl]: emacs.

File: src/channel_worker.erl [hunk 23]
```diff
@@ -75,7 +76,7 @@ stop(Channel) ->
 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
 %% @doc Looks up the pid for a given channel.
 
--spec where_is(ppspp_channel:channel()) -> {ok, pid()} | {error,_}.
+-spec where_is({channel, ppspp_channel:channel()}) -> {ok, pid()} | {error,_}.
```
Comment (elvisci) [inline src/channel_worker.erl]: Missing space after "," on line 79

File: src/ppspp_channel.erl [hunk 24]
```diff
@@ -130,7 +135,7 @@ where_is(Channel = {channel, _}) ->
 %% @end
 -spec get_swarm_id(channel()) ->
     {ok, ppspp_options:swarm_id() } | {error, any()}.
-get_swarm_id(Channel = {channel, _}) ->
+get_swarm_id(Channel) ->
     try gproc:lookup_value({n,l,Channel}) of
```
Comment (elvisci) [inline src/ppspp_channel.erl]: Missing space after "," on line 139

File: src/ppspp_handshake.erl [hunk 25]
```diff
@@ -89,7 +94,8 @@ handle({handshake, Body}, Endpoint) ->
         % prove useful to simply create a new channel anyway, which would
         % allow peers that dropped off for some reason to restart cleanly.
         false -> {ok, _Pid} =
-                 channel_worker:where_is(Inbound_channel)
+                 channel_worker:where_is({channel,Inbound_channel})
     end,
     % send handshake message to channel_worker for reply
-    channel_worker:handle(Pid, {handshake, Source_channel}).
+    ok = channel_worker:handle(Pid, #{handshake => #{channel => Source_channel}}),
```
Comment (elvisci) [inline src/ppspp_handshake.erl]: Line 100 is too long:     ok = channel_worker:handle(Pid, #{handshake => #{channel => Source_channel}}),.

File: src/ppspp_options.erl [hunk 26]
```diff
@@ -89,47 +91,47 @@
...
+          Maybe_Options/binary >>, Options) ->
     Method = case Maybe_Method of
                  0 -> no_integrity_protection;
                  1 -> merkle_hash_tree;
                  2 -> sign_all;
                  3 -> unified_merkle_hash_tree
              end,
-    Options = orddict:store(content_integrity_check_method, Method, Options0),
-    unpack(Maybe_Options, Options);
+    %% Options = orddict:store(content_integrity_check_method, Method, Options0),
```
Comment (elvisci) [inline src/ppspp_options.erl]: Line 130 is too long:     %% Options = orddict:store(content_integrity_check_method, Method, Options0),.

File: src/swirl.erl [hunk 27]
```diff
@@ -162,7 +162,7 @@ start_channel(Peer_endpoint, Swarm_options) ->
 
 %% @doc stop a PPSPP channel
 %% @end
--spec stop_channel(ppspp_channel:channel()) ->
+-spec stop_channel({channel,ppspp_channel:channel()}) ->
```
Comment (elvisci) [inline src/swirl.erl]: Missing space after "," on line 165

File: test/ppspp_have_SUITE.erl [hunk 28]
```diff
@@ -35,5 +35,6 @@ have(_Config) ->
     Method = chunk_32bit_chunks,
     Garbage = crypto:strong_rand_bytes(10),
     Chunks = <<15:64, Garbage/binary>>,
-    Expected = {{have,{chunk_spec,{chunk_32bit_chunks,0,15}}}, Garbage},
+    Expected = { #{have => #{chunk_spec => {chunk_32bit_chunks,0,15} }},
```
Comment (elvisci) [inline test/ppspp_have_SUITE.erl]: Missing space after "," on line 38

File: test/ppspp_options_SUITE.erl [hunk 29]
```diff
@@ -56,7 +57,7 @@ getters(_Config) ->
     Swarm_id = convert:hex_string_to_padded_binary(Root_Hash),
     Options = ppspp_options:use_default_options(Swarm_id),
     %% confirm that we can unwrap options from within a Datagram-style orddict
-    Options = ppspp_options:get_options([{foo, bar}, {fu, baa}, Options]),
+    Options = ppspp_options:get_options(#{foo => bar, fu => baa, options => Options}),
```
Comment (elvisci) [inline test/ppspp_options_SUITE.erl]: Line 60 is too long:     Options = ppspp_options:get_options(#{foo => bar, fu => baa, options => Options}),.

File: test/ppspp_options_SUITE.erl [hunk 30]
```diff
@@ -56,7 +57,7 @@ getters(_Config) ->
...
     % check returned values against header file defaults
     ?PPSPP_DEFAULT_CHUNK_METHOD = ppspp_options:get_chunk_addressing_method(Options),
```
Comment (elvisci) [inline test/ppspp_options_SUITE.erl]: Line 62 is too long:     ?PPSPP_DEFAULT_CHUNK_METHOD = ppspp_options:get_chunk_addressing_method(Options),.

File: test/ppspp_options_SUITE.erl [hunk 31]
```diff
@@ -56,7 +57,7 @@ getters(_Config) ->
...
     ?PPSPP_DEFAULT_INTEGRITY_METHOD = ppspp_options:get_content_integrity_check_method(Options),
```
Comment (elvisci) [inline test/ppspp_options_SUITE.erl]: Line 63 is too long:     ?PPSPP_DEFAULT_INTEGRITY_METHOD = ppspp_options:get_content_integrity_check_method(Options),.

**Comment by dch** [comment]
shaping up nicely! Do you think we are ready to merge now?

**Comment by kansi** [comment]
I think we shouldn't delay this much, we can fix bugs (if any) as we go.

**Comment by dch** [comment]
@kansi this needs to be rebased off /develop/ so we can get a clean merge, and the comments need to be removed. If I have time tonight I'll get to it.

File: src/channel_worker.erl [hunk 32]
```diff
@@ -65,7 +66,7 @@ start(Peer_endpoint, Swarm_options) ->
 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
 %% @doc Stops the server, given a channel.
 
--spec stop(ppspp_channel:channel()) -> ok | {error, any()}.
+-spec stop({channel, ppspp_channel:channel()}) -> ok | {error, any()}.
```
Comment (dch) [inline src/channel_worker.erl]: this doesn't feel right here -- either we're passing in an opaque object or we're not. Is there a reason not to use `#{channel => Channel}` as an opaque type from `ppspp_channel()` ?

File: src/channel_worker.erl [hunk 33]
```diff
@@ -109,7 +110,7 @@ start_link(Peer, Swarm_id)  ->
 %% address, and therefore multiple swarms on the same IP.
 %% TODO enable updating the registration of the URI to match the acquired
 %% channel that will accommodate all future datagrams from this peer.
--spec init(ppspp_options:swarm_id()) -> {ok, ppspp_channel:channel()}.
+-spec init(ppspp_options:swarm_id()) -> {ok, ppspp_channel:channel_ref()}.
```
Comment (dch) [inline src/channel_worker.erl]: We should be able to stick with `ppspp_channel:channel()` as before.

File: src/ppspp_channel.erl [hunk 34]
```diff
@@ -60,25 +61,29 @@ unpack_channel(Binary) ->
     Channel.
 
 -spec get_channel_id(channel()) -> non_neg_integer().
-get_channel_id(_Channel = {channel, Channel}) -> Channel.
+get_channel_id(Channel) -> Channel.
```
Comment (dch) [inline src/ppspp_channel.erl]: same again wrt keeping this consistent with orddict implementation.

File: src/ppspp_options.erl [hunk 35]
```diff
@@ -177,83 +181,91 @@ unpack( <<?PPSPP_LIVE_DISCARD_WINDOW, Maybe_Discard_Window/binary>>,
...
-    unpack(Maybe_Options, Options);
+          Maybe_Options/binary >>, Options) ->
+    Supported_Message_Types =
+        case Maybe_Supported_Message_Types of
+            %% TODO probably needs a custom bitmatching
+            %% parser
+            _ ->
+                ppspp_supported_messages_not_yet_implemented
+        end,
+    %% Options = orddict:store(ppspp_supported_message_types,
```
Comment (dch) [inline src/ppspp_options.erl]: per initial feedback, these comments need to go

**Comment by dch** [comment]
So I'm struggling to follow some of the changes here.

My reading of how the data structure should be implemented is as follows:

```erlang
Endpoint = #{
  channel => 0,
  ip => {127,0,0,1},
  port => 7777,
  socket => #Port<0.1727>,
  transport => udp,
  uri => "127.0.0.1:7777#0x00000000"}.

Options = #{
  chunk_addressing_method => chunking_32bit_chunks,
  chunk_size => 1024,
  content_integrity_check_method => merkle_hash_tree,
  merkle_hash_tree_function => sha256,
  minimum_version => 1,
  supported_version => 1,
  swarm_id => <<200,152,0,191,200,46,208,30,214,227,191,213,64,140,81,39,68,145,247,212>>}.

Datagram = #{
                endpoint => Endpoint,
                messages => [
                #{
                   handshake => #{
                                              channel => 19476988,
                                              options => Options}}]}.
%% again, but expanded out fully
Datagram = #{
    endpoint => #{
        channel => 0,
        ip => {127,0,0,1},
        port => 7777,
        socket => #Port<0.1727>,
        transport => udp,
        uri => "127.0.0.1:7777#0x00000000"},
  messages => [#{
        handshake => #{
            channel => 19476988,
            options => #{
                chunk_addressing_method => chunking_32bit_chunks,
                chunk_size => 1024,
                content_integrity_check_method => merkle_hash_tree,
                merkle_hash_tree_function => sha256,
                minimum_version => 1,
                supported_version => 1,
                swarm_id => <<200,152,0,191,200,46,208,30,214,227,191,213,64,140,81,39,68,145,247,212>>}}}]}.
```

Of note:

- datagrams comprise an endpoint (including the incoming channel id)
- a list of messages
- a handshake is embedded options and source channel id.

I can't square this manual conversion with the code. Any clarifications?

**Comment by kansi** [comment]
I think options and endpoint follow the mentioned structure. I didn't know datagram need to be just map and not {datagram, map()}. Will fix this issue today.

**Comment by dch** [comment]
yay! I'm out to lunch and paperwork, I'll pick it up again when I'm back & we should be merged today.

**Comment by igalic** [comment]
@dch thank you for this rebase. a couple of comments in general:

* there's no attribution to @kansi 
* the commit message is not very… good.

**Comment by dch** [comment]
this took a fair bit of time to sort out in the end. I've not checked the documentation build yet for the typespecs, but the map changes & tests are finally up to date.

**Comment by dch** [comment]
There are a few spots where I've used maps pattern matching that is not available in OTP 17.x, only in 18.x. I think we can drop support for non-current OTP releases and worry about this in The Future after a 1.0 release.

**Comment by dch** [comment]
docs don't work yet due to some breaking changes in `edown` that I need to review. I'd like to merge this to /develop/ anyway, and ensure @kansi gets credit in the merge commit.

**Comment by dch** [comment]
@kansi @igalic let me know if you're ok for me to push this to /develop/now?

**Comment by kansi** [comment]
re-checked the branch, LGTM.

**Pull request merged**